[ONNX][TypePromo] aten.div

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104491
* #104720
* __->__ #104229
* #104064

Unlike the majority of other operators, `aten.div` requires this specialized rule since its
type promotion kind depends on value of kwargs `rounding_mode`.

Checkout note `Update type promotion rule` in `type_promotion.py` before manually 
adding a type promotion rule.

BowenBao committed:
[ONNX][TypePromo] aten.div

[ghstack-poisoned]

thiagocrepaldi reviewed (commented):
Similar to https://github.com/pytorch/pytorch/pull/104063, please consider adding a description to the PR.

Can this PR be considered a good example on how to add a new type promotion rule? If so, this could be part of the description. It seems it does because it implements a new class which is "registered" at `_EXTRA_TYPE_PROMOTION_RULE_SET `

BowenBao committed:
Update rule docs and comments; Relax dispatcher check when producing multiple ops on "[ONNX][TypePromo] aten.div"


Specialized rule that depends on value of kwargs `rounding_mode`.
Checkout note `Update type promotion rule` in `type_promotion.py` before manually adding a type promotion rule.

[ghstack-poisoned]

justinchuby commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -171,6 +171,42 @@ def preview_type_promotion(
+        Rule depends on the value of the `rounding_mode` argument.
+        """
+        rounding_mode = kwargs.get("rounding_mode", None)
+        if rounding_mode is None:
+            # true_divide
+            self.promotion_kind = (
+                _prims_common.ELEMENTWISE_TYPE_PROMOTION_KIND.INT_TO_FLOAT
+            )
+            return super().preview_type_promotion(args, kwargs)
+        elif rounding_mode == "trunc":
+            # trunc_divide
+            self.promotion_kind = _prims_common.ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT
+            return super().preview_type_promotion(args, kwargs)
+        elif rounding_mode == "floor":
+            # floor_divide
+            self.promotion_kind = _prims_common.ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT
+            return super().preview_type_promotion(args, kwargs)
+        else:
+            raise ValueError(f"Unknown rounding_mode: {rounding_mode}")
```
```suggestion
        if rounding_mode is None:
            # true_divide
            self.promotion_kind = (
                _prims_common.ELEMENTWISE_TYPE_PROMOTION_KIND.INT_TO_FLOAT
            )
            return super().preview_type_promotion(args, kwargs)
        if rounding_mode == "trunc":
            # trunc_divide
            self.promotion_kind = _prims_common.ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT
            return super().preview_type_promotion(args, kwargs)
        if rounding_mode == "floor":
            # floor_divide
            self.promotion_kind = _prims_common.ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT
            return super().preview_type_promotion(args, kwargs)
        raise ValueError(f"Unknown rounding_mode: {rounding_mode}")
```

potentially

thiagocrepaldi commented on torch/onnx/_internal/fx/passes/type_promotion.py:
```diff
@@ -171,6 +171,42 @@ def preview_type_promotion(
+    def preview_type_promotion(
+        self, args: tuple, kwargs: dict
+    ) -> TypePromotionSnapshot:
+        """Reference type promotion rule from torch._refs.div.
+
+        Rule depends on the value of the `rounding_mode` argument.
```
That docstring would make more sense in the class docstring instead. It explains why that class specialization was needed in the first place :)

BowenBao committed:
Add comments & cleanup refactor on "[ONNX][TypePromo] aten.div"


Unlike the majority of other operators, `aten.div` requires this specialized rule since its
type promotion kind depends on value of kwargs `rounding_mode`.

Checkout note `Update type promotion rule` in `type_promotion.py` before manually 
adding a type promotion rule.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX][TypePromo] aten.div"


Unlike the majority of other operators, `aten.div` requires this specialized rule since its
type promotion kind depends on value of kwargs `rounding_mode`.

Checkout note `Update type promotion rule` in `type_promotion.py` before manually 
adding a type promotion rule.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX][TypePromo] aten.div"


Unlike the majority of other operators, `aten.div` requires this specialized rule since its
type promotion kind depends on value of kwargs `rounding_mode`.

Checkout note `Update type promotion rule` in `type_promotion.py` before manually 
adding a type promotion rule.

[ghstack-poisoned]

BowenBao committed:
rebase, CI failing left-right-center unrelated on "[ONNX][TypePromo] aten.div"


Unlike the majority of other operators, `aten.div` requires this specialized rule since its
type promotion kind depends on value of kwargs `rounding_mode`.

Checkout note `Update type promotion rule` in `type_promotion.py` before manually 
adding a type promotion rule.

[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge -f "unrelated cudnn"
